### ISSUE TITLE
Split up CustomAuthorizerEvent into separate interfaces for REQUEST and TOKEN

### ISSUE THREAD

**Issue by sgronblo** [DefinitelyTyped/DefinitelyTyped#24178]
If you know how to fix the issue, make a pull request instead.

- [X] I tried using the `@types/aws-lambda` package and had problems.
- [X] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript (Using typescript 2.6.1, however this is unrelated to tsc versions)
- [X] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [X] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Not sure who to tag since there are currently 22 contributors listed so I picked 5 of the most recent from the index.d.ts history please let me know if somebody should be tagged but isn't, Authors: @Haydabase @simonbuchan @andy-ms @coderbyheart @sheetalkamat 

I propose that the CustomAuthorizerEvent should be split up into two different types representing the REQUEST type custom authorizer and the TOKEN type custom authorizer. We just experienced a case where our custom authorizer Lambda was configured to be of type TOKEN but we were attempting to use the REQUEST attributes. Having this more explicit in the types would probably have made us realize this mistake earlier.

**Comment by simonbuchan** [comment]
So basically:

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5dc41ce74c24af9d569ca22e5a4277d574f18dc7/types/aws-lambda/index.d.ts#L64-L73

becomes, going by https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-custom-authorizer-input.html:

```ts
// API Gateway CustomAuthorizer "event"
export type CustomAuthorizerEvent =
  | CustomAuthorizerTokenEvent
  | CustomAuthorizerRequestEvent
  ;

export interface CustomAuthorizerTokenEvent {
    type: 'TOKEN';
    methodArn: string;
    authorizationToken: string;
    headers?: { [name: string]: string };
    pathParameters?: { [name: string]: string } | null;
    queryStringParameters?: { [name: string]: string } | null;
    requestContext?: APIGatewayEventRequestContext;
}

export interface CustomAuthorizerRequestEvent {
    type: 'REQUEST';
    methodArn: string;
    headers: { [name: string]: string };
    pathParameters: { [name: string]: string };
    queryStringParameters: { [name: string]: string };
    requestContext: APIGatewayEventRequestContext;
}
```

Seems pretty straightforward, but I'm concerned about:

> The following example shows an input to a REQUEST authorizer for an API method (GET /request) *with a proxy integration*:

Is that simply the `requestContext` parameter? And should it thus be parameterized, defaulting to any in case it's not using proxy integration?

**Comment by sgronblo** [comment]
Hmm, good question. One idea that comes to mind would be to make it parametric in this way

```
export type CustomAuthorizerEvent<T = any> =
  | CustomAuthorizerRequestEvent<T>
  | CustomAuthorizerTokenEvent

// CustomAuthorizerTokenEvent same as above

export interface CustomAuthorizerRequestEvent {
  type: 'REQUEST';
    methodArn: string;
    headers: { [name: string]: string };
    pathParameters: { [name: string]: string };
    queryStringParameters: { [name: string]: string };
    requestContext: APIGatewayEventRequestContext & T;
}
```

Hopefully even in the case of T being any editors would give nice autocompletion at least for the `APIGatewayEventRequestContext` properties but would allow any property to be accessed due to `& any`.

**Comment by simonbuchan** [comment]
OK, I'm going back through open `aws-lambda` issues since they seem to be being neglected. This seems like a good thing and I'll do a PR for it eventually. but it will be after #42080 and the more bug-fixey issues land.

I'm pretty sure this can be done (at least mostly) backwards compatible: just make sure that the alternatives define every property defined by any other, if only as `foo?: never` or similar, but it will need some decent tests of course.

**Comment by simonbuchan** [comment]
I must have some sort of memory issue for this event. I completely forgot about it when I noticed the issue testing a fix for another issue and opened #42420 for it.
